crates/sui/src/client_commands.rs

```diff
@@ -1217,6 +1217,9 @@ async fn compile_package(
             .into());
         }
     }
+    if with_unpublished_dependencies {
+        compiled_package.verify_unpublished_dependencies(&dependencies.unpublished)?;
+    }
```

Comment:
> Probably the neatest way to implement this is to change the way `with_unpublished_dependencies` is passed into `CompiledPackage::get_package_bytes`. Instead of just a boolean, have it accept a `&BTreeSet<Symbol>` representing the set of dependencies you want to include (which in our case would be `PackageDependencies::unpublished`), 

I tried this and ended up opting for a separate function on `compiled_package`. Passing  `&BTreeSet<Symbol>` (or `Option<&BTreeSet<Symbol>>`) seemed nice but involves a lot of updates to call sites/tests since the return value for `get_packages` becomes a `Result` and didn't seem worth it. Also that we'd want to retain a function that guaranteed success for getting module bytes, i.e., without a `Result` type, if there are no unpublished dependencies (empty `unpublished` or `None` set), which would probably mean minting a different function.

> (you might as well also check that all the modules in the root package have self-address set to `0x0` as well)

Right above this we have a block that I think does this. It's not integrated into a function (maybe it should be) but I think we're covered for checking this?

```rust
    if !compiled_package.is_framework() {
        if let Some(already_published) = compiled_package.published_root_module() {
            return Err(SuiError::ModulePublishFailure {
                error: format!(
                    "Modules must all have 0x0 as their addresses. \
                           Violated by module {:?}",
                    already_published.self_id(),
                ),
            }
            .into());
        }
    }
```